# GPU RF Shared-Book ML Trading

Train collapsed-label CUDA random forest classifiers per feature family, save reusable artifacts, and trade the out-of-sample predictions with one shared multi-asset Zipline book.

This is the classifier-only baseline for the classifier + latent autoencoder notebook. The notebook saves model metrics, score frames, trade logs, backtest summaries, and written analysis through `ArtifactStore`, so downstream notebooks load the actual latest baseline instead of copying numbers by hand.


In [1]:
from __future__ import annotations

from pathlib import Path
from time import perf_counter
import random
import sys
import warnings

import numpy as np
import pandas as pd

QUANT_WAREHOUSE_ROOT = Path('/home/jlee153232/PycharmProjects/quant-warehouse')
if str(QUANT_WAREHOUSE_ROOT) not in sys.path:
    sys.path.insert(0, str(QUANT_WAREHOUSE_ROOT))

import cupy as cp
import cudf

from quant_orchestrator.artifacts import ArtifactStore
from quant_orchestrator.platforms.backtesting_frameworks.shared_book import build_shared_book_weights
from quant_orchestrator.platforms.backtesting_frameworks.zipline.shared_book import (
    ZiplineSharedBookSummaryJob,
    run_zipline_shared_book_summary_jobs,
)
from quant_orchestrator.platforms.ml_frameworks.rapids import RapidsRandomForestClassifier
from quant_orchestrator.research_tools import (
    build_family_prediction_frame,
    build_strategy_score_frame,
    load_latest_experiment_artifacts,
    prepare_family_dataset,
    save_experiment_artifacts,
)

from quant_warehouse.platforms.data_providers.fmp.target_engineering.event_pairs import EventPairStore
from quant_warehouse.research_tools import (
    BinaryTargetConfig,
    FamilyEvaluationConfig,
    build_collapsed_bullish_event_target_panel,
    build_fundamental_feature_panel,
    build_oracle_trade_target_panel,
    cap_features_by_quality,
    load_fmp_event_pairs,
    screen_fmp_equity_universe,
)
from quant_warehouse.warehouse.api import Warehouse

pd.set_option('display.max_columns', 120)
pd.set_option('display.width', 180)
pd.set_option('display.float_format', lambda value: f'{value:,.4f}')
warnings.filterwarnings('ignore', category=FutureWarning)

print('cupy', cp.__version__, 'cuda_devices', cp.cuda.runtime.getDeviceCount())
print('cudf', cudf.__version__)


cupy 14.1.1 cuda_devices 1
cudf 26.06.00


In [2]:
RANDOM_SEED = 20260702
EXPERIMENT_NAME = 'gpu_rf_shared_book_100b'
MIN_MARKET_CAP = 100_000_000_000
START_DATE = '1900-01-01'
END_DATE = None
TRAIN_END = pd.Timestamp('2019-12-31')
OOS_START = pd.Timestamp('2020-01-01')

TOP_K_VALUES = [5, 10, 20, 40]
ENTRY_THRESHOLD = 0.50
EXIT_THRESHOLD = 0.50
MIN_FEATURE_COVERAGE = 0.50
MAX_FEATURES_PER_FAMILY = 50
MIN_TRAIN_ROWS_PER_FAMILY = 250
MIN_CLASSES_PER_FAMILY = 2
CAPITAL_BASE = 1_000_000.0

ZIPLINE_COMMISSION_PER_SHARE = 0.005
ZIPLINE_SLIPPAGE_BPS = 5.0
ZIPLINE_MAX_WORKERS = 4

RF_PARAMS = {
    'n_estimators': 300,
    'max_depth': 16,
    'max_features': 'sqrt',
    'n_bins': 128,
    'random_state': RANDOM_SEED,
    'n_streams': 8,
}

FEATURE_CONFIG = FamilyEvaluationConfig(
    provider='fmp',
    market_cap_min=MIN_MARKET_CAP,
    start_date=START_DATE,
    end_date=END_DATE,
    max_features_per_family=MAX_FEATURES_PER_FAMILY,
)
TARGET_CONFIG = BinaryTargetConfig(
    provider='fmp',
    start_date=START_DATE,
    end_date=END_DATE,
    event_families=('congress', 'insider', 'analyst_rating', 'price_target', 'earnings'),
    oracle_trade_k_by_frequency={'YE': tuple(range(1, 13))},
)

random.seed(RANDOM_SEED)
np.random.seed(RANDOM_SEED)
FEATURE_CONFIG, TARGET_CONFIG

(FamilyEvaluationConfig(provider='fmp', market_cap_min=100000000000, country='US', exchanges=('NASDAQ', 'NYSE', 'AMEX'), screen_limit=5000, start_date='1900-01-01', end_date=None, filing_lag_days=45, horizons=(20, 60, 120), min_observations=120, max_features_per_family=50),
 BinaryTargetConfig(provider='fmp', start_date='1900-01-01', end_date=None, event_families=('congress', 'insider', 'analyst_rating', 'price_target', 'earnings'), oracle_trade_k_by_frequency={'YE': (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)}, oracle_trade_min_profit_pct=0.01, oracle_trade_long_entry_price_col='high', oracle_trade_long_exit_price_col='low', oracle_trade_short_entry_price_col='low', oracle_trade_short_exit_price_col='high', event_alignment_tolerance_days=7, collapsed_bullish_event_types=('congress_buy', 'insider_buy', 'analyst_upgrade', 'price_target_raise', 'earnings_beat')))

## Build Data

Training uses every available row before 2020. Out-of-sample classification metrics and trading results both use 2020+ rows.

In [3]:
started = perf_counter()
WAREHOUSE = Warehouse()
EVENT_STORE = EventPairStore(backend=WAREHOUSE.backend, catalog=WAREHOUSE.catalog)

symbols, raw_universe, universe_eligibility, universe_source = screen_fmp_equity_universe(FEATURE_CONFIG, warehouse=WAREHOUSE)
print({'universe_source': universe_source, 'eligible_symbols': len(symbols)})
display(universe_eligibility.loc[universe_eligibility['eligible']].head(30))

{'universe_source': 'openbb:fmp', 'eligible_symbols': 117}


,symbol,eligible,reason,screen_market_cap
0,NVDA,True,ok,4785585180000
1,GOOGL,True,ok,4368788098535
2,GOOG,True,ok,4349493623926
3,AAPL,True,ok,4323663859280
4,MSFT,True,ok,2854597080400
5,AMZN,True,ok,2599991070000
6,SPCX,True,ok,2059971841733
7,AVGO,True,ok,1757164597200
8,TSLA,True,ok,1597307716000
9,META,True,ok,1555825021126


In [4]:
raw_feature_panel, raw_feature_metadata, feature_diagnostics, feature_timings = build_fundamental_feature_panel(
    symbols, FEATURE_CONFIG, warehouse=WAREHOUSE
)
selected_features, selected_feature_metadata, feature_quality = cap_features_by_quality(
    raw_feature_panel, raw_feature_metadata, max_features=MAX_FEATURES_PER_FAMILY
)
feature_panel = raw_feature_panel[['symbol', 'date', *selected_features]].copy()
feature_panel['symbol'] = feature_panel['symbol'].astype(str).str.upper()
feature_panel['date'] = pd.to_datetime(feature_panel['date'], errors='coerce').dt.normalize()
print({
    'raw_feature_panel_rows': len(raw_feature_panel),
    'selected_feature_columns': len(selected_features),
    'selected_feature_families': selected_feature_metadata['family'].nunique(),
    **feature_timings,
})
display(selected_feature_metadata.groupby(['source', 'family'], as_index=False).agg(feature_count=('feature', 'nunique')).sort_values(['source', 'family']))

{'raw_feature_panel_rows': 1155377, 'selected_feature_columns': 365, 'selected_feature_families': 15, 'raw_panel_build_seconds': 12.831504741217941}


,source,family,feature_count
0,financetoolkit,ft_growth_balance,50
1,financetoolkit,ft_growth_cash,50
2,financetoolkit,ft_growth_income,38
3,financetoolkit,ft_ratios_efficiency,5
4,financetoolkit,ft_ratios_liquidity,7
5,financetoolkit,ft_ratios_profitability,14
6,financetoolkit,ft_ratios_solvency,15
7,financetoolkit,ft_ratios_valuation,24
8,fmp,fmp_balance_mcap,50
9,fmp,fmp_cash_mcap,39


In [5]:
events, event_diagnostics, event_load_seconds = load_fmp_event_pairs(
    symbols, TARGET_CONFIG, event_store=EVENT_STORE, include_historical=True
)
event_symbols = tuple(event_diagnostics.loc[event_diagnostics['combined_rows'].gt(0), 'symbol'].sort_values())
feature_panel = feature_panel.loc[feature_panel['symbol'].isin(event_symbols)].copy()

collapsed_event_panel, collapsed_event_metadata = build_collapsed_bullish_event_target_panel(
    feature_panel[['symbol', 'date']], events, TARGET_CONFIG
)
oracle_panel, oracle_metadata, oracle_seconds = build_oracle_trade_target_panel(
    event_symbols, TARGET_CONFIG, warehouse=WAREHOUSE
)
print({
    'event_symbols': len(event_symbols),
    'event_rows': len(events),
    'event_load_seconds': round(event_load_seconds, 3),
    'collapsed_event_rows': len(collapsed_event_panel),
    'oracle_rows': len(oracle_panel),
    'oracle_columns': len(oracle_metadata),
    'oracle_seconds': round(oracle_seconds, 3),
})
display(event_diagnostics.sort_values('combined_rows', ascending=False).head(20))

{'event_symbols': 116, 'event_rows': 15212, 'event_load_seconds': 3.207, 'collapsed_event_rows': 1155366, 'oracle_rows': 1155366, 'oracle_columns': 36, 'oracle_seconds': 37.455}


,symbol,cached_rows,historical_rows,combined_rows,event_families
68,MSFT,65,971,1036,"(analyst_rating, congress, earnings, insider, ..."
0,AAPL,62,885,947,"(analyst_rating, congress, earnings, insider, ..."
73,NVDA,47,812,859,"(analyst_rating, congress, earnings, insider, ..."
7,AMZN,66,742,808,"(analyst_rating, congress, earnings, insider, ..."
43,GOOGL,60,590,650,"(analyst_rating, congress, earnings, insider, ..."
103,TSLA,60,459,519,"(analyst_rating, congress, earnings, insider, ..."
63,META,55,437,492,"(analyst_rating, congress, earnings, insider, ..."
57,LMT,0,132,132,"(earnings,)"
32,CVX,0,130,130,"(earnings,)"
80,PH,0,129,129,"(earnings,)"


In [6]:
def collapsed_label_rows(collapsed_event_panel: pd.DataFrame, oracle_panel: pd.DataFrame) -> tuple[pd.DataFrame, pd.DataFrame]:
    event_target = 'target_event_collapsed__bullish'
    event_activity = f'_target_activity__{event_target}'
    frames = []
    diagnostics = []

    event_frame = collapsed_event_panel.copy()
    if event_target in event_frame.columns:
        activity = pd.to_numeric(event_frame.get(event_activity, 0), errors='coerce').fillna(0).astype('int8')
        value = pd.to_numeric(event_frame[event_target], errors='coerce').fillna(0).astype('int8')
        bullish = event_frame.loc[value.eq(1), ['symbol', 'date']].copy()
        bullish['collapsed_label'] = 'event_bullish'
        bullish['label_source'] = 'event_collapsed'
        frames.append(bullish)
        diagnostics.append({'source': 'event_collapsed', 'candidate_rows': int(activity.gt(0).sum()), 'used_rows': len(bullish), 'dropped_rows': int(activity.gt(0).sum()) - len(bullish), 'note': 'mirror/non-bullish event rows excluded'})

    long_cols = sorted(c for c in oracle_panel.columns if c.startswith('target_oracle_trade_entry__') and c.endswith('_long'))
    short_cols = sorted(c for c in oracle_panel.columns if c.startswith('target_oracle_trade_entry__') and c.endswith('_short'))
    if long_cols and short_cols:
        oracle = oracle_panel[['symbol', 'date', *long_cols, *short_cols]].copy()
        long_any = oracle[long_cols].apply(pd.to_numeric, errors='coerce').fillna(0).gt(0).any(axis=1)
        short_any = oracle[short_cols].apply(pd.to_numeric, errors='coerce').fillna(0).gt(0).any(axis=1)
        ambiguous = long_any & short_any
        long_rows = oracle.loc[long_any & ~short_any, ['symbol', 'date']].copy()
        long_rows['collapsed_label'] = 'oracle_long'
        long_rows['label_source'] = 'oracle_trade'
        short_rows = oracle.loc[short_any & ~long_any, ['symbol', 'date']].copy()
        short_rows['collapsed_label'] = 'oracle_short'
        short_rows['label_source'] = 'oracle_trade'
        frames.extend([long_rows, short_rows])
        diagnostics.append({'source': 'oracle_trade', 'candidate_rows': int((long_any | short_any).sum()), 'used_rows': len(long_rows) + len(short_rows), 'dropped_rows': int(ambiguous.sum()), 'note': 'ambiguous long+short rows dropped after k collapse'})

    labels = pd.concat(frames, ignore_index=True) if frames else pd.DataFrame(columns=['symbol', 'date', 'collapsed_label', 'label_source'])
    labels['symbol'] = labels['symbol'].astype(str).str.upper()
    labels['date'] = pd.to_datetime(labels['date'], errors='coerce').dt.normalize()
    labels = labels.dropna(subset=['symbol', 'date', 'collapsed_label']).drop_duplicates()
    return labels.sort_values(['date', 'symbol', 'collapsed_label']).reset_index(drop=True), pd.DataFrame(diagnostics)

label_rows, label_diagnostics = collapsed_label_rows(collapsed_event_panel, oracle_panel)
print({'collapsed_label_rows': len(label_rows), 'classes': sorted(label_rows['collapsed_label'].unique())})
display(label_diagnostics)
display(label_rows.groupby(['label_source', 'collapsed_label'], as_index=False).agg(rows=('symbol', 'size'), symbols=('symbol', 'nunique'), min_date=('date', 'min'), max_date=('date', 'max')))

{'collapsed_label_rows': 115116, 'classes': ['event_bullish', 'oracle_long', 'oracle_short']}


,source,candidate_rows,used_rows,dropped_rows,note
0,event_collapsed,13507,10037,3470,mirror/non-bullish event rows excluded
1,oracle_trade,105079,105079,0,ambiguous long+short rows dropped after k coll...


,label_source,collapsed_label,rows,symbols,min_date,max_date
0,event_collapsed,event_bullish,10037,116,1993-01-08,2026-06-22
1,oracle_trade,oracle_long,53395,116,1962-01-02,2026-06-22
2,oracle_trade,oracle_short,51684,116,1962-01-02,2026-06-22


## Train GPU RF Family Models Pre-2020

Each feature family gets one CUDA random forest. Out-of-sample metrics are measured on 2020+ labeled rows.

In [7]:
model_rows = []
models = {}
train_started = perf_counter()
for source, family in selected_feature_metadata[['source', 'family']].drop_duplicates().sort_values(['source', 'family']).itertuples(index=False, name=None):
    family_frame, features = prepare_family_dataset(
        feature_panel,
        selected_feature_metadata,
        label_rows,
        source=str(source),
        family=str(family),
        min_feature_coverage=MIN_FEATURE_COVERAGE,
    )
    if family_frame.empty:
        model_rows.append({'source': source, 'family': family, 'status': 'skipped_empty', 'features': len(features), 'rows': 0})
        continue
    train = family_frame.loc[pd.to_datetime(family_frame['date']).le(TRAIN_END)].copy()
    oos = family_frame.loc[pd.to_datetime(family_frame['date']).ge(OOS_START)].copy()
    if len(train) < MIN_TRAIN_ROWS_PER_FAMILY or train['collapsed_label'].nunique() < MIN_CLASSES_PER_FAMILY:
        model_rows.append({'source': source, 'family': family, 'status': 'skipped_sparse_train', 'features': len(features), 'rows': len(family_frame), 'train_rows': len(train), 'oos_rows': len(oos), 'train_classes': train['collapsed_label'].nunique()})
        continue
    fit_started = perf_counter()
    classifier = RapidsRandomForestClassifier.fit(
        train,
        features=features,
        target_col='collapsed_label',
        random_state=RANDOM_SEED,
        params=RF_PARAMS,
    )
    fit_seconds = perf_counter() - fit_started
    models[(source, family)] = {'classifier': classifier, 'features': features}
    train_scores = classifier.score(train, features=features, target_col='collapsed_label')
    oos_scores = classifier.score(oos, features=features, target_col='collapsed_label') if not oos.empty and oos['collapsed_label'].nunique() > 1 else {'rows': len(oos), 'accuracy': np.nan, 'balanced_accuracy': np.nan, 'macro_f1': np.nan}
    model_rows.append({'source': source, 'family': family, 'status': 'ok', 'features': len(features), 'rows': len(family_frame), 'train_rows': len(train), 'oos_rows': len(oos), 'classes': family_frame['collapsed_label'].nunique(), 'fit_seconds': fit_seconds, **{f'train_{k}': v for k, v in train_scores.items()}, **{f'oos_{k}': v for k, v in oos_scores.items()}})

model_results = pd.DataFrame(model_rows).sort_values(['status', 'oos_macro_f1', 'oos_balanced_accuracy'], ascending=[True, False, False]).reset_index(drop=True)
print({'trained_models': len(models), 'elapsed_seconds': round(perf_counter() - train_started, 3), 'experiment_name': EXPERIMENT_NAME})
display(model_results)


[22:57:32] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:57:35] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:57:38] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:57:41] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:57:43] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:57:46] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:57:49] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:57:52] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:57:55] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:57:59] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:58:02] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:58:05] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:58:08] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


[22:58:11] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


{'trained_models': 15, 'elapsed_seconds': 46.678, 'experiment_name': 'gpu_rf_shared_book_100b'}


[22:58:15] /__w/nvforest/nvforest/python/nvforest/build/cp311-abi3-linux_aarch64/_deps/treelite-src/src/serializer.cc:202: The model you are loading originated from a newer Treelite version; some functionalities may be unavailable.
Currently running Treelite version 4.6.1
The model checkpoint was generated from Treelite version 4.7.0


,source,family,status,features,rows,train_rows,oos_rows,classes,fit_seconds,train_accuracy,train_balanced_accuracy,train_macro_f1,oos_accuracy,oos_balanced_accuracy,oos_macro_f1
0,financetoolkit,ft_ratios_valuation,ok,24,91991,71185,20806,3,2.0871,0.5190,0.3831,0.3710,0.4353,0.3930,0.3728
1,fmp,fmp_balance_mcap,ok,50,89847,69060,20787,3,2.2372,0.8392,0.6227,0.6061,0.4355,0.3584,0.3187
2,fmp,fmp_cash_mcap,ok,39,85269,64463,20806,3,2.2136,0.8292,0.6218,0.6054,0.4307,0.3542,0.3131
3,fmp,fmp_daily_mcap_yield,ok,14,89932,69126,20806,3,2.1340,0.8420,0.6227,0.6035,0.4295,0.3527,0.3117
4,fmp,fmp_daily_mcap_multiple,ok,14,89924,69118,20806,3,2.0919,0.8432,0.6251,0.6086,0.4293,0.3528,0.3109
5,fmp,fmp_income_mcap,ok,31,91166,70360,20806,3,2.1635,0.8373,0.6236,0.6113,0.4306,0.3562,0.3103
6,fmp,fmp_daily_ev_yield,ok,7,91887,71081,20806,3,2.0951,0.7800,0.5723,0.5510,0.4231,0.3474,0.3058
7,fmp,fmp_daily_ev_multiple,ok,7,91181,70375,20806,3,2.1146,0.7960,0.5870,0.5691,0.4218,0.3463,0.3052
8,financetoolkit,ft_ratios_efficiency,ok,5,91991,71185,20806,3,2.0389,0.5135,0.3770,0.3603,0.4162,0.3395,0.2881
9,financetoolkit,ft_ratios_solvency,ok,15,91991,71185,20806,3,2.1010,0.5164,0.3804,0.3662,0.4118,0.3341,0.2684


## Inference Scores

Scores are averaged across feature-family models. `long_score = P(event_bullish) + P(oracle_long)`, capped at 1. `short_score = P(oracle_short)`.

In [8]:
pred_frames = []
for (source, family), payload in models.items():
    pred_input = build_family_prediction_frame(feature_panel, payload['features'], min_feature_coverage=MIN_FEATURE_COVERAGE)
    if pred_input.empty:
        continue
    proba = payload['classifier'].predict_proba_frame(pred_input, payload['features'])
    scored = build_strategy_score_frame(
        source=str(source),
        family=str(family),
        prediction_frame=pred_input,
        probability_frame=proba,
        ae_familiarity_frame=None,
    )
    pred_frames.append(scored)

single_model_scores = pd.concat(pred_frames, ignore_index=True) if pred_frames else pd.DataFrame()
single_model_scores = single_model_scores.loc[pd.to_datetime(single_model_scores['date']).ge(OOS_START)].copy()

mean_scores = (
    single_model_scores.groupby(['symbol', 'date'], as_index=False)
    .agg(
        long_score=('long_score', 'mean'),
        short_score=('short_score', 'mean'),
        long_exit_score=('long_exit_score', 'mean'),
        short_exit_score=('short_exit_score', 'mean'),
        classifier_long_score=('classifier_long_score', 'mean'),
        classifier_short_score=('classifier_short_score', 'mean'),
        ae_familiarity=('ae_familiarity', 'mean'),
        ae_recon_error=('ae_recon_error', 'mean'),
        ae_latent_distance=('ae_latent_distance', 'mean'),
        model_count=('strategy_source', 'nunique'),
    )
)
mean_scores['source'] = 'ensemble'
mean_scores['family'] = 'mean'
mean_scores['strategy_source'] = 'ensemble_mean'
mean_scores['net_score'] = mean_scores['long_score'] - mean_scores['short_score']

score_cols = ['strategy_source', 'source', 'family', 'symbol', 'date', 'long_score', 'short_score', 'long_exit_score', 'short_exit_score', 'classifier_long_score', 'classifier_short_score', 'ae_familiarity', 'ae_recon_error', 'ae_latent_distance', 'net_score', 'model_count']
strategy_scores = pd.concat([mean_scores[score_cols], single_model_scores[score_cols]], ignore_index=True)

print({
    'ensemble_prediction_rows': len(mean_scores),
    'single_model_prediction_rows': len(single_model_scores),
    'strategy_sources': strategy_scores['strategy_source'].nunique(),
    'symbols': strategy_scores['symbol'].nunique(),
    'dates': strategy_scores['date'].nunique(),
})
display(mean_scores[['long_score', 'short_score', 'net_score', 'model_count']].describe())
display(
    strategy_scores.groupby(['strategy_source', 'source', 'family'], as_index=False)
    .agg(rows=('symbol', 'size'), symbols=('symbol', 'nunique'), dates=('date', 'nunique'))
    .sort_values(['strategy_source'])
)


{'ensemble_prediction_rows': 182606, 'single_model_prediction_rows': 2738961, 'strategy_sources': 16, 'symbols': 116, 'dates': 1627}


,long_score,short_score,net_score,model_count
count,"182,606.0000","182,606.0000","182,606.0000","182,606.0000"
mean,0.5544,0.4456,0.1087,14.9993
std,0.0287,0.0287,0.0574,0.0266
min,0.3486,0.2969,-0.3029,14.0000
25%,0.5403,0.4283,0.0806,15.0000
50%,0.5560,0.4440,0.1121,15.0000
75%,0.5717,0.4597,0.1435,15.0000
max,0.7031,0.6514,0.4063,15.0000


,strategy_source,source,family,rows,symbols,dates
0,ensemble_mean,ensemble,mean,182606,116,1627
1,financetoolkit.ft_growth_balance,financetoolkit,ft_growth_balance,182606,116,1627
2,financetoolkit.ft_growth_cash,financetoolkit,ft_growth_cash,182606,116,1627
3,financetoolkit.ft_growth_income,financetoolkit,ft_growth_income,182606,116,1627
4,financetoolkit.ft_ratios_efficiency,financetoolkit,ft_ratios_efficiency,182606,116,1627
5,financetoolkit.ft_ratios_liquidity,financetoolkit,ft_ratios_liquidity,182606,116,1627
6,financetoolkit.ft_ratios_profitability,financetoolkit,ft_ratios_profitability,182606,116,1627
7,financetoolkit.ft_ratios_solvency,financetoolkit,ft_ratios_solvency,182606,116,1627
8,financetoolkit.ft_ratios_valuation,financetoolkit,ft_ratios_valuation,182606,116,1627
9,fmp.fmp_balance_mcap,fmp,fmp_balance_mcap,182477,116,1627


In [9]:
def load_price_frames(symbols):
    frames = {}
    for symbol in symbols:
        prices = WAREHOUSE.read_prices(symbol, provider=FEATURE_CONFIG.provider, start=START_DATE, end=END_DATE)
        if prices is None or prices.empty:
            continue
        frame = prices.rename(columns=str.lower).copy()
        required = ['open', 'high', 'low', 'close', 'volume']
        if not set(required).issubset(frame.columns):
            continue
        frame = frame[required].apply(pd.to_numeric, errors='coerce')
        frame.index = pd.DatetimeIndex(pd.to_datetime(frame.index, errors='coerce')).normalize()
        frame = frame.dropna(subset=['open', 'high', 'low', 'close']).sort_index()
        if not frame.empty:
            frames[str(symbol).upper()] = frame
    return frames

price_frames = load_price_frames(event_symbols)
wide_close = pd.DataFrame({symbol: frame['close'] for symbol, frame in price_frames.items()}).sort_index().ffill()
next_returns = wide_close.pct_change().shift(-1)
print({'price_symbols': wide_close.shape[1], 'price_dates': wide_close.shape[0], 'oos_dates': int((wide_close.index >= OOS_START).sum())})
display(wide_close.tail())


{'price_symbols': 116, 'price_dates': 16252, 'oos_dates': 1627}


,AAPL,ABBV,ABT,ADI,AMAT,AMD,AMGN,AMZN,ANET,APH,APP,AVGO,AXP,BA,BAC,BKNG,BLK,BMY,BNY,BRK-A,BRK-B,BX,C,CAT,CDNS,COF,COP,COST,CRM,CRWD,CSCO,CVS,CVX,DE,DELL,DHR,DIS,FTNT,GE,GEV,GILD,GLW,GOOG,GOOGL,GS,HD,HONIV,HWM,IBKR,IBM,INTC,ISRG,JNJ,JPM,KLAC,KO,LLY,LMT,LOW,LRCX,MA,MCD,MDT,META,MO,MRK,MRVL,MS,MSFT,MU,NEE,NFLX,NOW,NVDA,ORCL,PANW,PEP,PFE,PG,PGR,PH,PLD,PLTR,PM,PNC,PWR,QCOM,RTX,SBUX,SCCO,SCHW,SNDK,SO,SOJE,SOMN,SPGI,SYK,T,TBB,TJX,TMO,TMUS,TSLA,TXN,UBER,UNH,UNP,V,VRT,VRTX,VZ,WDC,WELL,WFC,WMT,XOM
date,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
2026-06-17,295.9500,221.2300,88.5000,414.4500,592.9200,512.4800,341.6600,237.5000,164.9300,161.1100,479.4900,392.9000,340.5400,225.6300,56.5300,171.6300,"1,057.3800",55.2800,146.0700,"737,300.0000",491.2800,125.0100,143.7800,955.9200,389.6000,200.8700,111.2100,965.5900,155.0200,682.9600,117.3300,99.1600,177.5800,588.4700,419.3200,177.7600,100.8600,144.1400,357.0300,"1,048.8600",125.4500,175.4000,362.1000,363.7900,"1,099.1400",327.4800,190.0000,283.2300,95.0900,262.3500,121.1000,402.1800,234.2000,333.4600,238.7300,79.9300,"1,112.0000",532.3200,217.2700,374.1800,492.9900,283.8200,78.1400,567.5800,68.9500,115.4400,289.5400,224.9600,378.9100,"1,043.1900",85.7300,76.9600,95.4800,204.6500,183.5300,282.1300,141.5900,25.9200,150.5600,204.4200,945.9700,140.9400,130.6300,179.4400,233.1300,714.8500,212.9700,192.5800,99.8200,191.6800,94.5100,"1,958.8000",92.5300,16.9100,48.8200,418.0200,301.1400,22.4400,20.5900,164.1300,461.6900,181.3100,396.3800,301.8800,70.9100,399.5300,258.0500,330.3800,317.5800,458.9900,45.8400,712.1300,206.5500,83.8100,118.1300,140.7400
2026-06-18,298.0100,216.4900,88.4100,434.4600,617.1100,537.3700,337.6000,244.3900,169.6700,163.9600,469.7100,411.3500,338.0000,222.7200,56.2000,171.7800,"1,050.0900",54.0000,143.6300,"733,609.7700",489.4600,123.7900,143.0600,985.8200,387.3900,201.5300,107.7400,951.4500,151.7800,684.8600,119.5400,98.3200,173.6300,589.2400,409.5000,177.1700,103.8900,144.7300,357.6400,"1,109.7300",123.7600,194.9200,367.4600,368.0300,"1,096.5600",334.2800,206.0000,277.6600,96.0000,249.1000,133.9900,406.7800,228.3900,325.2200,259.5600,79.3900,"1,098.5700",510.9500,222.2000,389.0400,489.7900,278.6100,79.3400,577.2200,69.1200,113.8700,310.5800,223.1700,379.4000,"1,133.9900",86.7500,77.3800,95.0400,210.6900,184.2900,287.7800,142.0200,25.2100,150.3800,204.8700,953.2700,140.5400,128.4700,178.4000,232.0400,702.2500,226.1100,185.6000,100.6500,192.9300,91.7000,"2,184.7500",93.0900,16.9500,48.9200,410.9200,307.8000,22.0100,20.6000,163.8100,464.6100,181.6700,400.4900,322.8600,71.6400,400.9600,256.8800,327.2400,333.0500,451.6300,45.3700,746.2300,206.6500,82.2000,117.1800,137.8100
2026-06-22,297.0100,230.0100,87.8300,445.4800,640.1800,551.6300,344.7200,232.7900,174.5600,165.9600,469.3900,392.1300,338.0700,220.8300,57.3700,167.7700,"1,051.7400",54.7000,146.4400,"734,399.9900",488.6900,123.2600,145.6700,"1,022.2800",389.0400,200.7000,109.7000,951.3500,150.1200,675.4400,121.5300,101.3000,175.0600,598.5900,418.7100,178.1900,102.4500,145.3900,355.1200,"1,127.5900",124.8700,209.8300,348.7800,349.6800,"1,106.3700",326.6200,193.6600,280.3600,96.8200,252.2200,140.9400,402.9500,231.2900,331.4800,269.1600,79.5300,"1,102.0800",493.6000,214.4000,409.5400,484.0900,270.1000,79.2700,563.8500,69.5100,115.4800,307.8600,227.0900,367.3400,"1,211.3800",86.0800,72.8800,93.0100,208.6500,175.0700,286.4000,140.7100,25.0800,147.6800,207.3800,962.1400,143.8300,119.5000,173.1700,234.7100,740.1400,221.9000,181.8300,100.1500,189.9100,92.0300,"2,273.7300",93.4300,16.7900,48.7300,407.3900,304.6900,22.1000,20.5300,164.1700,464.0100,180.0600,405.0500,332.2800,71.4300,406.6800,259.9100,326.6000,357.9600,466.7900,45.3600,732.6200,211.4500,83.8400,117.1800,138.4700
2026-06-23,294.3000,234.7600,90.5300,407.2600,585.8800,519.8500,347.0100,234.1100,162.2000,158.7000,467.0200,380.1500,337.7800,216.7100,57.9100,168.9400,"1,015.3300",55.5300,147.2900,"737,80

## Shared-Book Trading Policy

For every day: exit held positions on the opposite score, then fill open slots from the best current candidates above the entry threshold. Long+short uses one shared `top_k` capacity across both sides.

In [10]:
oos_dates = pd.DatetimeIndex(sorted(set(strategy_scores['date']).intersection(next_returns.index)))
oos_dates = oos_dates[oos_dates >= OOS_START]
trade_symbols = tuple(sorted(set(strategy_scores['symbol']).intersection(next_returns.columns)))
print({'trade_symbols': len(trade_symbols), 'oos_dates': len(oos_dates), 'strategy_sources': strategy_scores['strategy_source'].nunique()})


{'trade_symbols': 116, 'oos_dates': 1627, 'strategy_sources': 16}


In [11]:
weight_artifacts = {}
trade_logs = []
zipline_jobs = []
price_window_start = pd.Timestamp(oos_dates.min()) - pd.Timedelta(days=20)
price_window_end = pd.Timestamp(oos_dates.max())
price_frame_subset = {
    symbol: frame.loc[(pd.DatetimeIndex(frame.index) >= price_window_start) & (pd.DatetimeIndex(frame.index) <= price_window_end)].copy()
    for symbol, frame in price_frames.items()
    if symbol in trade_symbols
}
strategy_source_order = ['ensemble_mean'] + sorted(
    source for source in strategy_scores['strategy_source'].dropna().unique().tolist() if source != 'ensemble_mean'
)

for strategy_source in strategy_source_order:
    score_frame = strategy_scores.loc[strategy_scores['strategy_source'].eq(strategy_source)].copy()
    if score_frame.empty:
        continue
    source_value = score_frame['source'].iloc[0]
    family_value = score_frame['family'].iloc[0]
    print({'strategy_source': strategy_source, 'rows': len(score_frame), 'symbols': score_frame['symbol'].nunique(), 'dates': score_frame['date'].nunique()})
    for variant in ['long_only', 'short_only', 'long_short']:
        for top_k in TOP_K_VALUES:
            weights, trades = build_shared_book_weights(
                score_frame,
                trade_symbols,
                oos_dates,
                top_k=top_k,
                variant=variant,
                entry_threshold=ENTRY_THRESHOLD,
                exit_threshold=EXIT_THRESHOLD,
                long_exit_score_col='long_exit_score',
                short_exit_score_col='short_exit_score',
            )
            weight_artifacts[(strategy_source, variant, top_k)] = weights
            trades = trades.assign(strategy_source=strategy_source, source=source_value, family=family_value, variant=variant, top_k=top_k)
            trade_logs.append(trades)
            zipline_jobs.append(
                ZiplineSharedBookSummaryJob(
                    price_frames=price_frame_subset,
                    target_weights=weights,
                    metadata={
                        'strategy_source': strategy_source,
                        'source': source_value,
                        'family': family_value,
                        'variant': variant,
                        'top_k': top_k,
                        'score_rows': len(score_frame),
                        'score_symbols': score_frame['symbol'].nunique(),
                        'score_dates': score_frame['date'].nunique(),
                        'signal_events': len(trades),
                        'commission_per_share': ZIPLINE_COMMISSION_PER_SHARE,
                        'slippage_bps': ZIPLINE_SLIPPAGE_BPS,
                        'avg_gross_exposure': float(weights.abs().sum(axis=1).mean()),
                        'avg_net_exposure': float(weights.sum(axis=1).mean()),
                        'fully_invested_days': float(weights.abs().sum(axis=1).ge(0.999).mean()),
                        'cash_days': float(weights.abs().sum(axis=1).eq(0).mean()),
                    },
                    capital_base=CAPITAL_BASE,
                    commission_per_share=ZIPLINE_COMMISSION_PER_SHARE,
                    slippage_bps=ZIPLINE_SLIPPAGE_BPS,
                )
            )

print({'zipline_jobs': len(zipline_jobs), 'zipline_max_workers': ZIPLINE_MAX_WORKERS})
backtest_summary = run_zipline_shared_book_summary_jobs(zipline_jobs, max_workers=ZIPLINE_MAX_WORKERS)
trade_log = pd.concat(trade_logs, ignore_index=True) if trade_logs else pd.DataFrame()
backtest_summary = backtest_summary.sort_values(['strategy_source', 'variant', 'top_k']).reset_index(drop=True)
metric_cols = ['framework', 'strategy_source', 'source', 'family', 'variant', 'top_k', 'trades', 'signal_events', 'final_equity', 'total_return', 'annualized_return', 'annualized_vol', 'sharpe', 'max_drawdown', 'avg_gross_exposure', 'avg_net_exposure', 'fully_invested_days', 'cash_days']
display(backtest_summary[metric_cols])
display(trade_log.groupby(['strategy_source', 'variant', 'top_k', 'action'], as_index=False).size().head(80))


{'strategy_source': 'ensemble_mean', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'financetoolkit.ft_growth_balance', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'financetoolkit.ft_growth_cash', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'financetoolkit.ft_growth_income', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'financetoolkit.ft_ratios_efficiency', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'financetoolkit.ft_ratios_liquidity', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'financetoolkit.ft_ratios_profitability', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'financetoolkit.ft_ratios_solvency', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'financetoolkit.ft_ratios_valuation', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'fmp.fmp_balance_mcap', 'rows': 182477, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'fmp.fmp_cash_mcap', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'fmp.fmp_daily_ev_multiple', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'fmp.fmp_daily_ev_yield', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'fmp.fmp_daily_mcap_multiple', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'fmp.fmp_daily_mcap_yield', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'strategy_source': 'fmp.fmp_income_mcap', 'rows': 182606, 'symbols': 116, 'dates': 1627}


{'zipline_jobs': 192, 'zipline_max_workers': 4}


,framework,strategy_source,source,family,variant,top_k,trades,signal_events,final_equity,total_return,annualized_return,annualized_vol,sharpe,max_drawdown,avg_gross_exposure,avg_net_exposure,fully_invested_days,cash_days
0,zipline_shared_book_native,ensemble_mean,ensemble,mean,long_only,5,6191,5,"2,072,636.6400",1.0726,0.1195,0.1882,0.6943,-0.2615,1.0000,1.0000,1.0000,0.0000
1,zipline_shared_book_native,ensemble_mean,ensemble,mean,long_only,10,13652,12,"3,585,938.9500",2.5859,0.2187,0.2025,1.0789,-0.2832,1.0000,1.0000,1.0000,0.0000
2,zipline_shared_book_native,ensemble_mean,ensemble,mean,long_only,20,25758,22,"3,126,540.5900",2.1265,0.1931,0.1841,1.0521,-0.2956,1.0000,1.0000,1.0000,0.0000
3,zipline_shared_book_native,ensemble_mean,ensemble,mean,long_only,40,46315,52,"2,898,650.8700",1.8987,0.1792,0.1890,0.9678,-0.3205,1.0000,1.0000,1.0000,0.0000
4,zipline_shared_book_native,ensemble_mean,ensemble,mean,long_short,5,6191,5,"2,072,636.6400",1.0726,0.1195,0.1882,0.6943,-0.2615,1.0000,1.0000,1.0000,0.0000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
187,zipline_shared_book_native,fmp.fmp_income_mcap,fmp,fmp_income_mcap,long_short,40,53082,470,"1,955,888.9900",0.9559,0.1095,0.1575,0.7392,-0.2689,1.0000,0.4555,1.0000,0.0000
188,zipline_shared_book_native,fmp.fmp_income_mcap,fmp,fmp_income_mcap,short_only,5,6985,63,"90,221.9800",-0.9098,-0.3110,0.3507,-0.8866,-0.9413,1.0000,-1.0000,1.0000,0.0000
189,zipline_shared_book_native,fmp.fmp_income_mcap,fmp,fmp_income_mcap,short_only,10,13478,210,"144,701.5300",-0.8553,-0.2587,0.3005,-0.8463,-0.8929,0.9980,-0.9980,0.9932,0.0000
190,zipline_shared_book_native,fmp.fmp_income_mcap,fmp,fmp_income_mcap,short_only,20,22869,642,"164,039.1200",-0.8360,-0.2442,0.2498,-0.9958,-0.8676,0.9858,-0.9858,0.9588,0.0000


,strategy_source,variant,top_k,action,size
0,ensemble_mean,long_only,5,enter_long,5
1,ensemble_mean,long_only,10,enter_long,11
2,ensemble_mean,long_only,10,exit_long,1
3,ensemble_mean,long_only,20,enter_long,21
4,ensemble_mean,long_only,20,exit_long,1
...,...,...,...,...,...
75,financetoolkit.ft_growth_income,long_short,5,exit_short,1
76,financetoolkit.ft_growth_income,long_short,10,enter_long,10
77,financetoolkit.ft_growth_income,long_short,10,enter_short,1
78,financetoolkit.ft_growth_income,long_short,10,exit_short,1


## Model vs Trading Comparison

In [12]:
model_oos_summary = model_results.loc[model_results['status'].eq('ok'), ['source', 'family', 'oos_rows', 'oos_accuracy', 'oos_balanced_accuracy', 'oos_macro_f1']].copy()
model_oos_summary['strategy_source'] = model_oos_summary.apply(lambda row: f'{row["source"]}.{row["family"]}', axis=1)
display(model_oos_summary.sort_values('oos_macro_f1', ascending=False))

strategy_comparison = (
    backtest_summary.loc[:, ['framework', 'strategy_source', 'source', 'family', 'variant', 'top_k', 'total_return', 'sharpe', 'max_drawdown', 'avg_gross_exposure', 'avg_net_exposure', 'trades']]
    .sort_values(['sharpe', 'total_return'], ascending=False)
    .reset_index(drop=True)
)
display(strategy_comparison)

best_by_strategy_source = (
    strategy_comparison.sort_values(['strategy_source', 'sharpe', 'total_return'], ascending=[True, False, False])
    .groupby('strategy_source', as_index=False)
    .head(1)
    .sort_values(['sharpe', 'total_return'], ascending=False)
    .reset_index(drop=True)
)
display(best_by_strategy_source)

model_vs_trading = best_by_strategy_source.merge(
    model_oos_summary[['strategy_source', 'oos_accuracy', 'oos_balanced_accuracy', 'oos_macro_f1']],
    on='strategy_source',
    how='left',
)
display(model_vs_trading.sort_values(['sharpe', 'total_return'], ascending=False))


,source,family,oos_rows,oos_accuracy,oos_balanced_accuracy,oos_macro_f1,strategy_source
0,financetoolkit,ft_ratios_valuation,20806,0.4353,0.3930,0.3728,financetoolkit.ft_ratios_valuation
1,fmp,fmp_balance_mcap,20787,0.4355,0.3584,0.3187,fmp.fmp_balance_mcap
2,fmp,fmp_cash_mcap,20806,0.4307,0.3542,0.3131,fmp.fmp_cash_mcap
3,fmp,fmp_daily_mcap_yield,20806,0.4295,0.3527,0.3117,fmp.fmp_daily_mcap_yield
4,fmp,fmp_daily_mcap_multiple,20806,0.4293,0.3528,0.3109,fmp.fmp_daily_mcap_multiple
5,fmp,fmp_income_mcap,20806,0.4306,0.3562,0.3103,fmp.fmp_income_mcap
6,fmp,fmp_daily_ev_yield,20806,0.4231,0.3474,0.3058,fmp.fmp_daily_ev_yield
7,fmp,fmp_daily_ev_multiple,20806,0.4218,0.3463,0.3052,fmp.fmp_daily_ev_multiple
8,financetoolkit,ft_ratios_efficiency,20806,0.4162,0.3395,0.2881,financetoolkit.ft_ratios_efficiency
9,financetoolkit,ft_ratios_solvency,20806,0.4118,0.3341,0.2684,financetoolkit.ft_ratios_solvency


,framework,strategy_source,source,family,variant,top_k,total_return,sharpe,max_drawdown,avg_gross_exposure,avg_net_exposure,trades
0,zipline_shared_book_native,fmp.fmp_daily_ev_yield,fmp,fmp_daily_ev_yield,long_only,20,3.6131,1.2513,-0.3121,1.0000,1.0000,30083
1,zipline_shared_book_native,fmp.fmp_daily_mcap_multiple,fmp,fmp_daily_mcap_multiple,long_short,5,4.3610,1.2498,-0.3557,1.0000,0.5621,7865
2,zipline_shared_book_native,fmp.fmp_income_mcap,fmp,fmp_income_mcap,long_only,5,5.3713,1.2317,-0.3418,1.0000,1.0000,7984
3,zipline_shared_book_native,fmp.fmp_daily_ev_multiple,fmp,fmp_daily_ev_multiple,long_only,10,3.8886,1.2142,-0.2999,1.0000,1.0000,15720
4,zipline_shared_book_native,fmp.fmp_balance_mcap,fmp,fmp_balance_mcap,long_only,40,3.0408,1.2086,-0.3259,1.0000,1.0000,54753
...,...,...,...,...,...,...,...,...,...,...,...,...
187,zipline_shared_book_native,fmp.fmp_daily_mcap_yield,fmp,fmp_daily_mcap_yield,short_only,40,-0.7481,-1.2709,-0.7753,0.7097,-0.7097,30207
188,zipline_shared_book_native,fmp.fmp_daily_mcap_multiple,fmp,fmp_daily_mcap_multiple,short_only,40,-0.7972,-1.3471,-0.8192,0.7973,-0.7973,33201
189,zipline_shared_book_native,fmp.fmp_daily_ev_yield,fmp,fmp_daily_ev_yield,short_only,20,-0.9039,-1.4481,-0.9198,0.9397,-0.9397,23932
190,zipline_shared_book_native,fmp.fmp_daily_ev_multiple,fmp,fmp_daily_ev_multiple,short_only,40,-0.7915,-1.4567,-0.8208,0.7536,-0.7536,35467


,framework,strategy_source,source,family,variant,top_k,total_return,sharpe,max_drawdown,avg_gross_exposure,avg_net_exposure,trades
0,zipline_shared_book_native,fmp.fmp_daily_ev_yield,fmp,fmp_daily_ev_yield,long_only,20,3.6131,1.2513,-0.3121,1.0000,1.0000,30083
1,zipline_shared_book_native,fmp.fmp_daily_mcap_multiple,fmp,fmp_daily_mcap_multiple,long_short,5,4.3610,1.2498,-0.3557,1.0000,0.5621,7865
2,zipline_shared_book_native,fmp.fmp_income_mcap,fmp,fmp_income_mcap,long_only,5,5.3713,1.2317,-0.3418,1.0000,1.0000,7984
3,zipline_shared_book_native,fmp.fmp_daily_ev_multiple,fmp,fmp_daily_ev_multiple,long_only,10,3.8886,1.2142,-0.2999,1.0000,1.0000,15720
4,zipline_shared_book_native,fmp.fmp_balance_mcap,fmp,fmp_balance_mcap,long_only,40,3.0408,1.2086,-0.3259,1.0000,1.0000,54753
5,zipline_shared_book_native,fmp.fmp_cash_mcap,fmp,fmp_cash_mcap,long_only,5,5.6995,1.1833,-0.3631,1.0000,1.0000,7938
6,zipline_shared_book_native,fmp.fmp_daily_mcap_yield,fmp,fmp_daily_mcap_yield,long_only,20,2.7979,1.1788,-0.3196,1.0000,1.0000,29783
7,zipline_shared_book_native,financetoolkit.ft_growth_balance,financetoolkit,ft_growth_balance,long_short,40,2.5587,1.0859,-0.3159,1.0000,0.9966,46442
8,zipline_shared_book_native,financetoolkit.ft_ratios_efficiency,financetoolkit,ft_ratios_efficiency,long_only,5,3.4266,1.0858,-0.3265,1.0000,1.0000,7879
9,zipline_shared_book_native,ensemble_mean,ensemble,mean,long_only,10,2.5859,1.0789,-0.2832,1.0000,1.0000,13652


,framework,strategy_source,source,family,variant,top_k,total_return,sharpe,max_drawdown,avg_gross_exposure,avg_net_exposure,trades,oos_accuracy,oos_balanced_accuracy,oos_macro_f1
0,zipline_shared_book_native,fmp.fmp_daily_ev_yield,fmp,fmp_daily_ev_yield,long_only,20,3.6131,1.2513,-0.3121,1.0000,1.0000,30083,0.4231,0.3474,0.3058
1,zipline_shared_book_native,fmp.fmp_daily_mcap_multiple,fmp,fmp_daily_mcap_multiple,long_short,5,4.3610,1.2498,-0.3557,1.0000,0.5621,7865,0.4293,0.3528,0.3109
2,zipline_shared_book_native,fmp.fmp_income_mcap,fmp,fmp_income_mcap,long_only,5,5.3713,1.2317,-0.3418,1.0000,1.0000,7984,0.4306,0.3562,0.3103
3,zipline_shared_book_native,fmp.fmp_daily_ev_multiple,fmp,fmp_daily_ev_multiple,long_only,10,3.8886,1.2142,-0.2999,1.0000,1.0000,15720,0.4218,0.3463,0.3052
4,zipline_shared_book_native,fmp.fmp_balance_mcap,fmp,fmp_balance_mcap,long_only,40,3.0408,1.2086,-0.3259,1.0000,1.0000,54753,0.4355,0.3584,0.3187
5,zipline_shared_book_native,fmp.fmp_cash_mcap,fmp,fmp_cash_mcap,long_only,5,5.6995,1.1833,-0.3631,1.0000,1.0000,7938,0.4307,0.3542,0.3131
6,zipline_shared_book_native,fmp.fmp_daily_mcap_yield,fmp,fmp_daily_mcap_yield,long_only,20,2.7979,1.1788,-0.3196,1.0000,1.0000,29783,0.4295,0.3527,0.3117
7,zipline_shared_book_native,financetoolkit.ft_growth_balance,financetoolkit,ft_growth_balance,long_short,40,2.5587,1.0859,-0.3159,1.0000,0.9966,46442,0.4124,0.3337,0.2492
8,zipline_shared_book_native,financetoolkit.ft_ratios_efficiency,financetoolkit,ft_ratios_efficiency,long_only,5,3.4266,1.0858,-0.3265,1.0000,1.0000,7879,0.4162,0.3395,0.2881
9,zipline_shared_book_native,ensemble_mean,ensemble,mean,long_only,10,2.5859,1.0789,-0.2832,1.0000,1.0000,13652,NaN,NaN,NaN


In [13]:
analysis_lines = [
    '## Written Analysis',
    '',
    f'- Experiment artifact name: `{EXPERIMENT_NAME}`.',
    f'- Universe: {len(symbols)} FMP 100B+ symbols; {len(event_symbols)} had event coverage.',
    f'- Training window: all available rows through {TRAIN_END.date()}. Out-of-sample model and trading window starts {OOS_START.date()}.',
    f'- Trained GPU RF feature-family models: {len(models)}.',
    f'- Strategy sources traded: {strategy_scores["strategy_source"].nunique()} total, including the ensemble mean and {max(strategy_scores["strategy_source"].nunique() - 1, 0)} individual feature-family models.',
    f'- Ensemble OOS prediction rows: {len(mean_scores):,} across {mean_scores["symbol"].nunique()} symbols and {mean_scores["date"].nunique()} dates.',
    f'- Strategy variants: long_only, short_only, long_short with top_k={TOP_K_VALUES}; trade size is fixed at 1/top_k and unused slots stay cash.',
    f'- Execution: native Zipline multi-asset shared-book engine using order_target_percent, ${ZIPLINE_COMMISSION_PER_SHARE:.3f}/share commission, and {ZIPLINE_SLIPPAGE_BPS:.1f} bps slippage.',
]
if not model_oos_summary.empty:
    best_model = model_oos_summary.sort_values('oos_macro_f1', ascending=False).iloc[0]
    analysis_lines.append(f'- Best OOS classifier family: {best_model["strategy_source"]} with macro_f1={best_model["oos_macro_f1"]:.4f}.')
if not backtest_summary.empty:
    best_trade = backtest_summary.sort_values(['sharpe', 'total_return'], ascending=False).iloc[0]
    ensemble_best = backtest_summary.loc[backtest_summary['strategy_source'].eq('ensemble_mean')].sort_values(['sharpe', 'total_return'], ascending=False).head(1)
    single_best = backtest_summary.loc[~backtest_summary['strategy_source'].eq('ensemble_mean')].sort_values(['sharpe', 'total_return'], ascending=False).head(1)
    analysis_lines.extend([
        '',
        'Best native Zipline trading row by Sharpe:',
        f'- {best_trade["strategy_source"]} / {best_trade["variant"]} top_k={int(best_trade["top_k"])}: total_return={best_trade["total_return"]:.2%}, sharpe={best_trade["sharpe"]:.2f}, max_drawdown={best_trade["max_drawdown"]:.2%}, avg_gross={best_trade["avg_gross_exposure"]:.2f}, avg_net={best_trade["avg_net_exposure"]:.2f}.',
    ])
    if not ensemble_best.empty:
        row = ensemble_best.iloc[0]
        analysis_lines.append(f'- Best ensemble_mean row: {row["variant"]} top_k={int(row["top_k"])} with total_return={row["total_return"]:.2%}, sharpe={row["sharpe"]:.2f}, max_drawdown={row["max_drawdown"]:.2%}.')
    if not single_best.empty:
        row = single_best.iloc[0]
        analysis_lines.append(f'- Best individual-model row: {row["strategy_source"]} / {row["variant"]} top_k={int(row["top_k"])} with total_return={row["total_return"]:.2%}, sharpe={row["sharpe"]:.2f}, max_drawdown={row["max_drawdown"]:.2%}.')
    analysis_lines.extend([
        '',
        'Interpretation:',
        '- This run is the artifact-backed classifier-only baseline for the 100B+ universe.',
        '- The per-model strategy variants show whether one feature family is carrying the ensemble or whether averaging adds value.',
        '- Compare model_vs_trading to see whether OOS classifier metrics line up with trading Sharpe; they often will not because ranking, thresholding, and exposure control are separate from multiclass accuracy.',
        '- This notebook uses a native execution engine for the shared multi-asset book instead of single-symbol sleeve runners or cost-only framework rows.',
    ])
analysis_markdown = '\n'.join(analysis_lines)
artifact_result = save_experiment_artifacts(
    experiment_name=EXPERIMENT_NAME,
    params={
        'min_market_cap': MIN_MARKET_CAP,
        'train_end': str(TRAIN_END.date()),
        'oos_start': str(OOS_START.date()),
        'top_k_values': TOP_K_VALUES,
        'entry_threshold': ENTRY_THRESHOLD,
        'exit_threshold': EXIT_THRESHOLD,
        'rf_params': RF_PARAMS,
    },
    metrics={
        'symbols': int(len(symbols)),
        'event_symbols': int(len(event_symbols)),
        'trained_models': int(len(models)),
        'strategy_sources': int(strategy_scores['strategy_source'].nunique()),
        'best_sharpe': None if backtest_summary.empty else float(backtest_summary['sharpe'].max()),
        'best_total_return': None if backtest_summary.empty else float(backtest_summary.sort_values(['sharpe', 'total_return'], ascending=False).iloc[0]['total_return']),
    },
    model_results=model_results,
    strategy_scores=strategy_scores,
    backtest_summary=backtest_summary,
    trade_log=trade_log,
    analysis_markdown=analysis_markdown,
)
print({'artifact_run_id': artifact_result.run.id, 'artifact_uris': artifact_result.artifact_uris})
from IPython.display import Markdown, display
display(Markdown(analysis_markdown))


{'artifact_run_id': 'run_639396c145be4d3399644ad8ebf73bc1', 'artifact_uris': {'model_results': 'artifact:art_2203f34daef64cdaa052c6b5ed0af4e0', 'strategy_scores': 'artifact:art_62c23f6048f2475c8dfc614183dc41b4', 'backtest_summary': 'artifact:art_c0602752339342e49cb30a373a0856dd', 'trade_log': 'artifact:art_f9abf48bc72c4d33ae927d5abfb25698', 'analysis': 'artifact:art_9867e9082ec84be59418209870bb99c5'}}


## Written Analysis

- Experiment artifact name: `gpu_rf_shared_book_100b`.
- Universe: 117 FMP 100B+ symbols; 116 had event coverage.
- Training window: all available rows through 2019-12-31. Out-of-sample model and trading window starts 2020-01-01.
- Trained GPU RF feature-family models: 15.
- Strategy sources traded: 16 total, including the ensemble mean and 15 individual feature-family models.
- Ensemble OOS prediction rows: 182,606 across 116 symbols and 1627 dates.
- Strategy variants: long_only, short_only, long_short with top_k=[5, 10, 20, 40]; trade size is fixed at 1/top_k and unused slots stay cash.
- Execution: native Zipline multi-asset shared-book engine using order_target_percent, $0.005/share commission, and 5.0 bps slippage.
- Best OOS classifier family: financetoolkit.ft_ratios_valuation with macro_f1=0.3728.

Best native Zipline trading row by Sharpe:
- fmp.fmp_daily_ev_yield / long_only top_k=20: total_return=361.31%, sharpe=1.25, max_drawdown=-31.21%, avg_gross=1.00, avg_net=1.00.
- Best ensemble_mean row: long_only top_k=10 with total_return=258.59%, sharpe=1.08, max_drawdown=-28.32%.
- Best individual-model row: fmp.fmp_daily_ev_yield / long_only top_k=20 with total_return=361.31%, sharpe=1.25, max_drawdown=-31.21%.

Interpretation:
- This run is the artifact-backed classifier-only baseline for the 100B+ universe.
- The per-model strategy variants show whether one feature family is carrying the ensemble or whether averaging adds value.
- Compare model_vs_trading to see whether OOS classifier metrics line up with trading Sharpe; they often will not because ranking, thresholding, and exposure control are separate from multiclass accuracy.
- This notebook uses a native execution engine for the shared multi-asset book instead of single-symbol sleeve runners or cost-only framework rows.

## Written Analysis

- Experiment artifact name: `gpu_rf_shared_book_100b`.
- Universe: 117 FMP 100B+ symbols; 116 had event coverage.
- Training window: all available rows through 2019-12-31. Out-of-sample model and trading window starts 2020-01-01.
- Trained GPU RF feature-family models: 15.
- Strategy sources traded: 16 total, including the ensemble mean and 15 individual feature-family models.
- Ensemble OOS prediction rows: 182,606 across 116 symbols and 1627 dates.
- Strategy variants: long_only, short_only, long_short with top_k=[5, 10, 20, 40]; trade size is fixed at 1/top_k and unused slots stay cash.
- Execution: native Zipline multi-asset shared-book engine using order_target_percent, $0.005/share commission, and 5.0 bps slippage.
- Best OOS classifier family: financetoolkit.ft_ratios_valuation with macro_f1=0.3728.

Best native Zipline trading row by Sharpe:
- fmp.fmp_daily_ev_yield / long_only top_k=20: total_return=361.31%, sharpe=1.25, max_drawdown=-31.21%, avg_gross=1.00, avg_net=1.00.
- Best ensemble_mean row: long_only top_k=10 with total_return=258.59%, sharpe=1.08, max_drawdown=-28.32%.
- Best individual-model row: fmp.fmp_daily_ev_yield / long_only top_k=20 with total_return=361.31%, sharpe=1.25, max_drawdown=-31.21%.

Interpretation:
- This run is the artifact-backed classifier-only baseline for the 100B+ universe.
- The per-model strategy variants show whether one feature family is carrying the ensemble or whether averaging adds value.
- Compare model_vs_trading to see whether OOS classifier metrics line up with trading Sharpe; they often will not because ranking, thresholding, and exposure control are separate from multiclass accuracy.
- This notebook uses a native execution engine for the shared multi-asset book instead of single-symbol sleeve runners or cost-only framework rows.